Mathlib/Algebra/AddTorsor.lean

```diff
@@ -0,0 +1,497 @@
+-- vsub_vadd' p₁ p₂ := show (p₁.1 -ᵥ p₂.1 +ᵥ p₂.1, _) = p₁ by simp
+--   ⊢ (p₁.fst -ᵥ p₂.fst +ᵥ p₂.fst, ((p₁.fst -ᵥ p₂.fst, p₁.snd -ᵥ p₂.snd) +ᵥ p₂).snd) = p₁
+-- vadd_vsub' v p := show (v.1 +ᵥ p.1 -ᵥ p.1, v.2 +ᵥ p.2 -ᵥ p.2) = v by simp
+--   ⊢ (v.fst +ᵥ p.fst -ᵥ p.fst, v.snd) = v
+
+@[simp]
+theorem fst_vadd (v : G × G') (p : P × P') : (v +ᵥ p).1 = v.1 +ᵥ p.1 :=
+  rfl
+#align prod.fst_vadd Prod.fst_vadd
+
+@[simp]
+theorem snd_vadd (v : G × G') (p : P × P') : (v +ᵥ p).2 = v.2 +ᵥ p.2 :=
+  rfl
+#align prod.snd_vadd Prod.snd_vadd
+
+@[simp]
+theorem mk_vadd_mk (v : G) (v' : G') (p : P) (p' : P') : (v, v') +ᵥ (p, p') = (v +ᵥ p, v' +ᵥ p') :=
+  rfl
+#align prod.mk_vadd_mk Prod.mk_vadd_mk
+
+@[simp]
+theorem fst_vsub (p₁ p₂ : P × P') : (p₁ -ᵥ p₂ : G × G').1 = p₁.1 -ᵥ p₂.1 :=
+  rfl
+#align prod.fst_vsub Prod.fst_vsub
+
+@[simp]
+theorem snd_vsub (p₁ p₂ : P × P') : (p₁ -ᵥ p₂ : G × G').2 = p₁.2 -ᵥ p₂.2 :=
+  rfl
+#align prod.snd_vsub Prod.snd_vsub
+
+@[simp]
+theorem mk_vsub_mk (p₁ p₂ : P) (p₁' p₂' : P') :
+    ((p₁, p₁') -ᵥ (p₂, p₂') : G × G') = (p₁ -ᵥ p₂, p₁' -ᵥ p₂') :=
+  rfl
+#align prod.mk_vsub_mk Prod.mk_vsub_mk
+
+end Prod
+
+namespace Pi
+
+universe u v w
+
+variable {I : Type u} {fg : I → Type v} [∀ i, AddGroup (fg i)] {fp : I → Type w}
+
+open AddAction AddTorsor
+
+/-- A product of `add_torsor`s is an `add_torsor`. -/
+instance [T : ∀ i, AddTorsor (fg i) (fp i)] : AddTorsor (∀ i, fg i) (∀ i, fp i) where
+  vadd g p i := g i +ᵥ p i
+  zero_vadd p := funext fun i => zero_vadd (fg i) (p i)
+  add_vadd g₁ g₂ p := funext fun i => add_vadd (g₁ i) (g₂ i) (p i)
+  vsub p₁ p₂ i := p₁ i -ᵥ p₂ i
+  Nonempty := ⟨fun i => Classical.choice (T i).Nonempty⟩
+  vsub_vadd' p₁ p₂ := funext fun i => vsub_vadd (p₁ i) (p₂ i)
+  vadd_vsub' g p := funext fun i => vadd_vsub (g i) (p i)
+
+end Pi
+
+namespace Equiv
+
+variable {G : Type _} {P : Type _} [AddGroup G] [AddTorsor G P]
+
+-- Porting note: Removed:
+-- include G
+
+/-- `v ↦ v +ᵥ p` as an equivalence. -/
+def vaddConst (p : P) : G ≃ P where
+  toFun v := v +ᵥ p
+  invFun p' := p' -ᵥ p
+  left_inv _ := vadd_vsub _ _
+  right_inv _ := vsub_vadd _ _
+#align equiv.vadd_const Equiv.vaddConst
+
+@[simp]
+theorem coe_vadd_const (p : P) : ⇑(vaddConst p) = fun v => v +ᵥ p :=
+  rfl
+#align equiv.coe_vadd_const Equiv.coe_vadd_const
+
+@[simp]
+theorem coe_vadd_const_symm (p : P) : ⇑(vaddConst p).symm = fun p' => p' -ᵥ p :=
+  rfl
+#align equiv.coe_vadd_const_symm Equiv.coe_vadd_const_symm
+
+/-- `p' ↦ p -ᵥ p'` as an equivalence. -/
+def constVsub (p : P) : P ≃ G where
+  toFun := (· -ᵥ ·) p
+  invFun v := -v +ᵥ p
+  left_inv p' := by simp
+  right_inv v := by simp [vsub_vadd_eq_vsub_sub]
+#align equiv.const_vsub Equiv.constVsub
+
+@[simp]
+theorem coe_const_vsub (p : P) : ⇑(constVsub p) = (· -ᵥ ·) p :=
+  rfl
+#align equiv.coe_const_vsub Equiv.coe_const_vsub
+
+@[simp]
+theorem coe_const_vsub_symm (p : P) : ⇑(constVsub p).symm = fun v => -v +ᵥ p :=
```

Comment:
```suggestion
theorem coe_const_vsub_symm (p : P) : (constVsub p).symm = (fun (v : G) => (-v +ᵥ p : P)) :=
```

putting all the type annotations seems to help this thing to elaborate, but I guess this isn't the big issue in this file